File: node/Node.py
class Node(object):

	def __init__(self, name):
		self.name = name
		self.children = []
		self.value = None

	def get_child(self, name):
		for child in self.children:
			if child.name == name:
				return child

	def add_child(self, name):
		if self.get_child(name) != None:
			return False
		self.children.append(Node(name))
		return True

	def get_value(self):
		return self.value

	def set_value(self, val):
		self.value = val

	def __str__(self):
		return "[Node#{}]".format(self.name)

class NodeNavigator(object):

	def __init__(self, root_node, delimiter):
		self.root_node = root_node
		self.delim = delimiter

	def _nav(self, paths, create):
		current_node = self.root_node
		for path in paths:
			if create:
				current_node.add_child(path)
			current_node = current_node.get_child(path)
			if current_node == None:
				raise Exception("while navigating through node {}, child with name {} does not exist".format(current_node, path))
		return current_node

	def get_node(self, path):
		return self._nav(path.split(self.delim), False)

	def create_folders(self, path):
		return self._nav(path.split(self.delim), True)

	def get_value(self, path):
		return self._nav(path.split(self.delim), False).get_value()

	def set_value(self, path, val):
		self._nav(path.split(self.delim), True).set_value(val)
		

# Testing starts here

mn = Node("MainNode")
nn = NodeNavigator(mn, "/")

while True:
	raw_args = input("=> ").split(" ")
	if raw_args[0] == "set":
		nn.set_value(raw_args[1], raw_args[2])
	elif raw_args[0] == "get":
		print(nn.get_value(raw_args[1]))
	else:
		print("?"+raw_args[0])

File: node/node.rb

class Node
  
  def initialize(name)
    @name = name
    @children = Array.new
    @value = nil
  end

  def set_value(value)
    @value = value
  end

  def get_value
    return @value
  end
  
  def get_child(name)
    ret_val = nil
    @children.each do |child|
      if child.instance_variable_get(:@name) == name
        ret_val = child
        break
      end
    end
    return ret_val
  end

  def add_child(name)
    if self.get_child(name) != nil
      return false
    end
    @children << Node.new(name)
    return true
  end

  def to_s
    "[Node##{@name}]"
  end
  
end

class NodeNavigator

  def initialize(root_node, delim)
    @root_node = root_node
    @delimiter = delim
  end

  def get_node_by_path(path)
    navigate(path.split(@delimiter), false)
  end

  def create_nodes(path)
    navigate(path.split(@delimiter), true)
  end

  def get_value(path)
    navigate(path.split(@delimiter), false).get_value
  end

  def set_value(path, val)
    navigate(path.split(@delimiter), true).set_value(val)
  end

  private
  def navigate(paths, create)
    current_node = @root_node
    paths.each do |path|
      if create then
        current_node.add_child(path)
      end
      current_node = current_node.get_child(path)
      if current_node == nil then
        raise "While navigating through node #{current_node}, next node #{path} does not exist"
      end
    end
    current_node
  end
end

# testing only

main_node = Node.new("main-node")
main_nav = NodeNavigator.new(main_node, '/')

while true
  print "@> "
  input = gets.split(' ')
  
  if input[0] == "set"
    main_nav.set_value(input[1], input[2])
    puts "value set"
  elsif input[0] == "get"
    puts main_nav.get_value(input[1])
  else
    puts "?#{input[0]}"
  end
end
